### DISCUSSION TITLE
제페토 테스트 시 캐릭터 생성이 안되는 문제

### DISCUSSION THREAD

**Discussion by LeeYongSu1** [naverz/zepeto-studio-kor#135] [5. Q&A]
안녕하세요 
캐릭터 로드 시 스크립트를 붙이고 초기화 하는 중에 GetComponent함수를 사용해서 스크립트 가져오기를 시도하면 캐릭터 생성이 안되는 문제가 있습니다.  유니티에서 플레이해서 테스트 하면 잘 실행이 되는데 제페토 어플로 테스트를 하면 생성이 되지 않습니다.

스크립트를 붙이는 부분입니다.

![image](https://user-images.githubusercontent.com/80555875/147213227-876679ff-1093-48b0-ba62-5e75889bbff3.png)

붙인 스크립트 초기화 하는 스크립트입니다.
빨간 줄 코드가GetComponent로 스크립트 가져오는 구문입니다. 
주석처리하면 캐릭터가 생성이 되고, 주석을 풀면 캐릭터 생성이 되지 않습니다.
 가져온 UI_PlayerController 스크립트에 문제가 있나 싶어 전부 주석으로 바꾸고 빈 스크립트로 바꿔도 같은 현상이 지속됩니다.

![image](https://user-images.githubusercontent.com/80555875/147212324-1d796808-3d87-4476-aab8-b7f61cb92086.png)

추가적으로   **
**this.playerUI = GameObject.Find("UI_CarController").transform.GetComponentInChildren<UI_PlayerController>();
                   this.playerUI = GameObject.Find("UI_CarController").transform.GetComponentInChildren(UI_PlayerController) as UI_PlayerController;  
                   this.playerUI = GameObject.FindGameObjectWithTag("asdf").transform.GetComponent<UI_PlayerController>();****
이런식으로 가져오는 방법을 바꿔도 같습니다. 
어떤 문제로 캐릭터 생성이 되지 않는지 도움을 주시면 감사하겠습니다.

**Comment by moondory77-naverz** [comment]
안녕하세요.
OnAddedLocalPlayer callback 이벤트는, 로컬캐릭터가 Scene에 완전히 로드가 완료된 이후 호출됩니다.

우선 해당 이벤트 타이밍에, 커스텀 스크립트 AddComponent에는  문제가 없음을 확인했습니다.

PlayerCtrl 컴포넌트를 추가하는 부분에서, 의도와 다른 상황이 나타난다면
일단 import하려는 PlayerCtrl의 로직에서부터 단계적으로 접근하실 필요가 있을 듯 합니다.

커스텀 스크립트를 import하는 부분도 다시한번 참고해주시면 좋을 듯 합니다. 
https://studio.zepeto.me/kr/guides/script-import

감사합니다.

**Reply by LeeYongSu1** [reply]
빠른 댓글 감사 드립니다.
문제를 찾기 위해서 PlayerCtrl에서 UI_PlayerController를 GetComponent하는 부분에 다른 빈 새로운 스크립트를 가져오기 시도해도 같은 문제가 발생합니다. 구문 하나 하나 주석을 넣었다 풀면서 문제를 확인했는데 해결이 되지 않네요... 어떤 부분을 보여드리면 도움을 주시기 편하실까요??

**Comment by moondory77-naverz** [comment]
안녕하세요. 
말씀주신 타이밍에, 빈 스크립트를 addComponent 할 때에도 문제가 발생한다면
커스텀 스크립트를 통해 Import 하는 방법에 문제가 있을 가능성이 있어 보입니다.

가장 좋은 방법은 공개 샘플 예제를 통해
작동 여부와, 어떤 부분에 차이가 있는지 확인해보는 것일 듯 합니다.
[샘플예제](https://github.com/naverz/zepeto-script-sample)

![addComponent](https://user-images.githubusercontent.com/93751484/147429030-479cb604-8a16-41b3-a351-def1e49856e3.png)

다음은 공개 샘플 예제의 10.ScriptImport에 작성되어있는 ExtraComponent를
2.Advanced/CharacterController/CharacterControlSample.ts 에서 AddComponent 하는 과정입니다.

동일한 OnAddedLocalPlayer 타이밍에, 커스텀스크립트를 addComponent하는 과정을 테스트 할 수 있습니다.
현재 문제가 발생하고 계시는 부분에 도움이 될 수 있는지 확인 부탁드리겠습니다.

**Reply by LeeYongSu1** [reply]
OnAddedLocalPlayer에 커스텀스크립트를 붙이는 부분은 정상 작동이 되고 있습니다. 다만, 붙인 커스텀스크립트에서 특정 오브젝트가 가지고 있는 스크립트를 GetComponent하는 구문이 들어가면 정상 작동이 되지 않습니다.

OnAddedLocalPlayer 타이밍에 커스텀스크립트Add -> _**붙인 커스텀스크립트 runtime 타이밍에 오브젝트 찾아서 GetComponet<다른 커스텀스크립트 or 빈 스크립트>()를 하는 과정**_입니다.

이 과정(붙인 커스텀스크립트 runtime 타이밍)에 자기 자신의 컴포넌트를 가져오거나, 다른 오브젝트에 접근해서 gameObject나 Transform을 가져오는 부분을 잘 동작을 하는데 스크립트를 GetComponent하면 동작을 하지 않습니다. 

<예제 ClientStarter에서 커스텀 스크립트 Add>
![image](https://user-images.githubusercontent.com/80555875/147630548-5c89ff2f-0813-4245-8a7d-7b8f23f38823.png)

<Add된 커스텀스크립트에서 다른 오브젝트 접근 후 스크립트 Get>
![image](https://user-images.githubusercontent.com/80555875/147630590-41610e9e-960d-474f-b447-8c5704b9b758.png)

<Get하는 스크립트>
![image](https://user-images.githubusercontent.com/80555875/147630617-ad41db4c-0d47-4fa1-b39f-229fead44399.png)


동적으로 붙여준 스크립트에서 다른 스크립트에 접근하려고 하면 문제가 발생한다고 생각되는데 확인 부탁드려도 될까요?
